### ISSUE TITLE
What is currently used as build evironment?

### ISSUE THREAD

**Issue by ronny332** [emukidid/swiss-gc#549]
Hi Team!
As I created a very slim FTP server for the Classic XBox years ago, it would be nice to port it over to Swiss, or design a new one. I spent the evening in digging into the build system of Swiss, combined with a lot of stumbling.
Besides a lot of other unsolvable issues (undeclared function stubs, missing includes, wrong version numbers for instance for libogc), I ask myself: "how do you build Swiss in a fast and convenient way?"
Building takes minutes, updating already compiled code seems not to work out of the box.

My linux box is a normal Intel based Ubuntu (64bit) System. I used the binary builds of Devkitpro for Debian, and for Arch in a simple VM.

**Comment by Extrems** [comment]
You need to build and install our libogc fork from https://github.com/emukidid/libogc or https://github.com/extremscorner/libogc2.

**Comment by ronny332** [comment]
Just because I already opened this topic and don't was to pollute the devkitpro's forum and this issues tracker with a new question:

Is there more documentation for how to organize code with devkitpro? The way include files are separated from their source files feels strange, but I got it working. What's a big issue, is that the header files are included again and again an the include  guards are not working, caused by the single compilation of each .c file.
Don't get me wrong, I not now to C/C++, but I don't want to hack the build system if there is a way how to make it right without further modifications.
Taking looks to the Swiss code makes it strange even more. For instance the variables of bba.h are declared extern, there is no obvious connection between the .c and .h file.

**Comment by Extrems** [comment]
I'm not sure what you're doing, but don't use the top-level makefile for development. Go at the level you're working on.

**Comment by ronny332** [comment]
Let's take these two files from the Swiss source code:

bba.h
```
  1 #ifndef __BBA_H
  2 #define __BBA_H
  3 
  4 extern char bba_ip[16];
  5 extern int net_initialized;
  6 extern int bba_exists;
  7 
  8 void init_network();
  9 
 10 #endif
```

bba.c
```
  1 #include <string.h>
  2 #include <unistd.h>
  3 #include <malloc.h>
  4 #include <network.h>
  5 #include <ogcsys.h>
  6 #include "exi.h"
  7 
  8 /* Network Globals */
  9 int net_initialized = 0;
 10 int bba_exists = 0;
 11 char bba_ip[16];
 12 
 13 // Init the GC net interface (bba)
 14 void init_network(void *args) {
 15 
 16     int res = 0;
 17 
 18     bba_exists = exi_bba_exists();
 19     if(bba_exists && !net_initialized) {
 20         res = if_config(bba_ip, NULL, NULL, true, 20);
 21         if(res >= 0 && strcmp("255.255.255.255", bba_ip)) {
 22             net_initialized = 1;
 23         }
 24         else {
 25             memset(bba_ip, 0, sizeof(bba_ip));
 26             net_initialized = 0;
 27         }
 28     }
 29 }
 ```
The .c file does not include the .h file and includes all needed headers by itself. This is fully different from the way normally .c and .h files are used.
If this is the "right" way for gamecube/devkitpro coding, I will be fine with it. But if not, I want to stop here and use the better way ;-).

**Comment by Extrems** [comment]
It isn't the right way. Swiss is just written like garbage.

**Comment by emukidid** [comment]
> It isn't the right way. Swiss is just written like garbage.

Positive criticism appreciated as always :D 

@ronny332 you're more than welcome to produce a PR with header/includes changes please if they are bothering you. It's mostly that a "this works, leave it alone" approach was taken in the early days when the project didn't have much interest and it was just me working on it. I wasn't exactly writing it to any standards.

**Comment by ronny332** [comment]
@emukidid no problem at all :blush:. What I saw until now makes a lot of fun for upcoming tests and tries to get a (lets call it) "drop in" working. My normal work is meanwhile far away from writing so memory reduced and (sometimes) self repeating code, which at the end does not really rely on many compiler optimizations, like we write code today in C(++).
"garbage" code from a more or less one person project hasn't to be bad at all. It is close to impossible to write always the same code style over years without concrete guideline, and who in person creates them whole other things have to be done? When I took a look at my first bigger software projects, starting at the end of the 90s, it often makes me smile, but at the end, it got the work done over years :blush:.
